File: discipline4/app/src/main/java/com/example/app/vaults/VaultData.kt
package com.example.app

import com.example.app.*
import androidx.room.Entity

/**
 * Represents vault data content with length constraints
 */
@Entity
@JvmInline
public value class VaultData private constructor(val value: String) {
  
  companion object {
    const val MINIMUM_LENGTH = 1
    const val MAXIMUM_LENGTH = 500
    
    /**
     * Creates VaultData with validation
     */
    fun create(string: String): Tried<VaultData, TextualError> {
      if (string.length < MINIMUM_LENGTH || string.length > MAXIMUM_LENGTH) {
        return Tried.failure(
          TextualError.create("Creating VaultData from string")
            .addMessage("String violates length invariants")
            .addNumberAttachment("Minimum length", MINIMUM_LENGTH.toDouble())
            .addNumberAttachment("Maximum length", MAXIMUM_LENGTH.toDouble())
            .addNumberAttachment("Provided string length", string.length.toDouble())
            .addStringAttachment("Provided string", string)
        )
      }
      
      return Tried.success(VaultData(string))
    }
    
    /**
     * Creates VaultData without validation (use with caution)
     */
    fun construct(string: String): VaultData {
      return VaultData(string)
    }
  }
    
  fun isEqualTo(other: VaultData): Boolean {
    return value == other.value
  }
  
  fun length(): Int {
    return value.length
  }
  
  fun isEmpty(): Boolean {
    return value.isEmpty()
  }

  override fun toString(): String {
    return "VaultData($value)"
  }
}

File: discipline_ts/core/vault_data.ts
import { TextualError, Unique } from "../x.ts";

const BRAND = Symbol();

export type VaultData = Unique<typeof BRAND, "VaultData", string>;

const MINIMUM_LENGTH = 1;
const MAXIMUM_LENGTH = 500;

const construct = (string: string): VaultData => {
  return string as VaultData;
};

const wrapOrThrow = (string: string): VaultData => {
  if (
    string.length < MINIMUM_LENGTH 
    || 
    string.length > MAXIMUM_LENGTH
  ) {
    const textualError = TextualError.create("Creating VaultData from string");
    TextualError.addMessage(textualError, "String violates length invariants. ");
    TextualError.addNumberAttachment(textualError, "Minimum length", MINIMUM_LENGTH);
    TextualError.addNumberAttachment(textualError, "Maximum length", MAXIMUM_LENGTH);
    TextualError.addNumberAttachment(textualError, "Provided string's length", string.length);
    TextualError.addStringAttachment(textualError, "Provided string", string);
    throw TextualError.toJsError(textualError);
  }

  return construct(string);
};

const getValue = (it: VaultData): string => {
  return it;
};

const isEqualTo = (it: VaultData, rhs: VaultData): boolean => {
  return it === rhs;
};

export const VaultData = {
  MINIMUM_LENGTH,
  MAXIMUM_LENGTH,
  wrapOrThrow,
  construct,
  isEqualTo,
  getValue,
};